Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClient.cs

```diff
@@ -19,10 +21,40 @@ namespace BuildXL.Cache.ContentStore.Service.Grpc
     /// An implementation of a CAS copy helper client based on GRPC.
     /// TODO: Consolidate with GrpcClient to deduplicate code. (bug 1365340)
     /// </summary>
-    public class GrpcCopyClient : IShutdown<BoolResult>
+    public sealed class GrpcCopyClient : IShutdown<BoolResult>
     {
+        private static ConcurrentDictionary<(string, int), GrpcCopyClient> _clientPool = new ConcurrentDictionary<(string, int), GrpcCopyClient>();
+
+        /// <summary>
+        /// Use an existing GRPC client if possible, else create a new one.
+        /// </summary>
+        /// <param name="host">Name of the host for the server (e.g. 'localhost').</param>
+        /// <param name="grpcPort">GRPC port on the server.</param>
+        public static GrpcCopyClient Create(string host, int grpcPort)
+        {
+            Console.WriteLine($"Creating GRPC client for {host},{grpcPort}");
+            if (_clientPool.TryGetValue((host, grpcPort), out var existingClient))

```

Comment:
To be honest, I would prefer no caching at all instead of having a cache that is effectively a memory leak (a cache without invalidation policy).

Lack of caching means that the app is just slow. But the current version can cause some super weird errors because of port exhaustion or because of some other issues..

Reply:
Agree in principle, but if you look closely this is keyed off of host and port. Perhaps we should consider making the GrpcCopyClient a weak reference so it has a chance to be collected, and we can reinstate when needed.

---
In reply to: [271135869](https://github.com/Microsoft/BuildXL/pull/66#discussion_r271135869) [](ancestors = 271135869)

Reply:
Just noticed that `ShutDownAsync` removes it. So if we have a guarantee that even in error cases and timeouts and cancellation that shutdown method is called, this should be no worse than all the other caches we have created. I'd still love to see at some point all these caches reported in a single object that we can measure holistically for memory and usage with well known lifetimes.

---
In reply to: [271395644](https://github.com/Microsoft/BuildXL/pull/66#discussion_r271395644) [](ancestors = 271395644,271135869)